### PR TITLE
Implement auth_tls api

### PR DESCRIPTION
**Pull request by kaelumania**
Implements: https://github.com/hashicorp/vault-ruby/issues/23

I was unable to come up with integration specs for the auth part, because the vault dev server does not support tls. Further I didn't implement the`crl` and `config` functionality for the `auth/cert/` path.

### CODE REVIEW

File: lib/vault/api/auth_tls.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    def auth_tls
+      @auth_tls ||= AuthTLS.new(self)
+    end
+  end
+
+  class AuthTLS < Request
+
+    # This enables the auth tls cert backend
+    def enable
+      client.sys.enable_auth('cert', 'cert', 'Allow to login via TLS certificate')
```
Comment (djmaze) [inline lib/vault/api/auth_tls.rb]: - [ ] `Allow login`

File: lib/vault/api/auth_tls.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    #   Vault.auth_tls.certificates #=> ["web"]
+    #
+    # @return [Array<String>]
+    def certificates
+      json = client.get("/v1/auth/cert/certs", list: true)
+      json[:data][:keys] || []
+    rescue HTTPError => e
+      return [] if e.code == 404
+      raise
+      []
```
Comment (djmaze) [inline lib/vault/api/auth_tls.rb]: - [ ] Superfluous

File: lib/vault/api/auth_tls.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    # Saves a certificate with the given name and attributes.
+    #
+    # @example
+    #   cert = Certificate.new('web-cert', '-----BEGIN CERTIFICATE...', "default", 3600)
+    #   Vault.auth_tls.put_certificate("web", cert) #=> true
+    #
+    # @param [String] name
+    #   the name of the certificate
+    # @param [Vault::Certificate] certificate
+    #   the certificate defintion
```
Comment (djmaze) [inline lib/vault/api/auth_tls.rb]: - [ ] definition

File: spec/support/sample_certificate.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        qL1oOzqiVnjuHIXekBbQJxJj886lbuR+mDSTo+sI79bQJ45W01NzHqAZ96VcS+cY
+        18Y5deKATxFaSDx8EBB+l38JCMnYBKSIMl7lHswBgjlNyL/fKC9dFlYTWdGycIOg
+        n+WiIBECgYATmx5+Uqtyz20WpSAY5QD2A+/Jnsth3YAYThYCfoUE0dJrWxI1fmUs
+        4dFt+QScfSPIWNAkPjeQinXerflrze/3FkmZ+X21fao33Ymq+RvT+O8Wsz6FmMR0
+        wyEvPHSgjDQ0jEwyRd8PhhZa13Zj2XoteoBwTRea20WSKetbUxI88A==
+        -----END RSA PRIVATE KEY-----
+      EOH
+    end
+  end
+end
```
Comment (djmaze) [inline spec/support/sample_certificate.rb]: I'd rather use fixture files for these.

Comment (sethvargo) [inline spec/support/sample_certificate.rb]: I actually like having everything inline. Fixtures are too far separated from the tests IMO.

**Comment by kaelumania** [comment]
@sethvargo ping

**Comment by kaelumania** [comment]
@sethvargo i would be nice to get some feedback on this branch. yes, travis fails with ruby 1.9, but i am not sure what the reason is.

File: lib/vault/api/auth_tls.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    #
+    # @return [true]
+    def put_certificate(name, certificate)
+      client.post("/v1/auth/cert/certs/#{CGI.escape(name)}", JSON.fast_generate(certificate.to_h))
```
Comment (amiryal) [inline lib/vault/api/auth_tls.rb]: In Ruby 1.9.3, instances of `Struct` don’t have `#to_h`; you may try to use `Hash[certificate.each_pair.to_a]` for compatibility instead.

**Comment by kaelumania** [comment]
@amiryal thanks, but merging the current master and adjusting the response class helped a lot. @sethvargo what do you think about the `to_h` method in the response class?

**Comment by sethvargo** [comment]
https://github.com/hashicorp/vault-ruby/pull/94

**Pull request closed without merging**